### ISSUE TITLE
Training with imbalanced S3DIS-like data set

### ISSUE THREAD

**Issue by dsheacanopy** [HuguesTHOMAS/KPConv-PyTorch#2]
I am attempting to train a model on custom point cloud data in a similar format to the S3DIS data set. One of these classes is only minimally represented relative to the size of the point cloud as a whole, and any training results in a model that simply never predicts that class.

Are there any configuration changes that I can make to prevent this from happening? In particular, I see that SemanticKitti has a `balanced_class=True` parameter, but S3DIS does not appear to have any such equivalent. I tried setting `segloss_balance = 'class'` in `train_S3DIS.py`, but that did not appear to make a difference.

Apologies in advance if this is not the appropriate place to ask this. I am essentially curious whether there is any class balancing for S3DIS data sets. Thanks!

**Comment by HuguesTHOMAS** [comment]
Hi @dsheacanopy,

This is an interesting question. There are many ways to help minority classes. In my current implementation, you have two options (that you can use separately or combine).

The first is using the option `use_potentials=False` when creating the S3DISDataset. For S3DIS, I have designed two dataloader strategies. 
 * `use_potentials=True` : this strategy uses values on every points that I call potentials, when I pick an input sphere in the dataset, I update the potential value by incrementing them. And the next sphere will be picked at the minimum of potentials. This strategy ensure a geometric consistency in the sphere picking. We pick every part of the point cloud the same amount of time, compared to a random picking, where in practice, some part of the point cloud could actually never be picked, especially if they have low density.
* `use_potentials=False`: This strategy is similar to the `balanced_class=True` in SemanticKitti.py. It is a random picking of the input sphere, but balanced by class. If the epoch need N sphere, we choose N/C sphere centered on a point of each class (where C is the number of classes). This strategy ensures that the network sees the minority class more often.
The second strategy is the one you are looking for, but remember it will be affected by varying densities as it remains a random picking.

The second option for you is to use the parameter `class_w`, which controls the weight of each class in the loss. There is an example where I try using that here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/ccb820bbb7a18e4642c317bc8a80d26164f7f3c1/train_SemanticKitti.py#L186-L188
If you use the first option, you can find a function to compute the class proportion as seen by your loss here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/ccb820bbb7a18e4642c317bc8a80d26164f7f3c1/datasets/SemanticKitti.py#L1420

Don't hesitate to share your results here. I am curious to know if you manage to predict your minority class and how. In my case, class_w had nearly no effect, and the first option was the one which help the most.

Best,
Hugues

**Comment by dsheacanopy** [comment]
Thank you so much for your help! I was able to get it to make predictions on the minority class by using both approaches simultaneously. I did have some success with each approach by itself, but it appears that both together seem to have produced the best improvement in results.

I will close this issue since the core problem is resolved, but I did have to make a couple small changes on my side to get it to run. In particular:

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/datasets/S3DIS.py#L513

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/utils/trainer.py#L587

These were raising an IndexError when setting `use_potentials=False` so I ended up just breaking on the exception. I don't know if this is the preferred workaround, or if that should have been happening at all, but it appears to have worked for me.

Thanks again!
